Userland/Libraries/LibGL/Clipper.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021, Jesse Buhagiar <jooster669@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/ScopeGuard.h>
+#include <LibGL/Clipper.h>
+
+bool GL::Clipper::point_within_clip_plane(const FloatVector4& vertex, ClipPlane plane)
+{
+    switch (plane) {
+    case ClipPlane::LEFT:
+        return vertex.x() > -vertex.w();
+    case ClipPlane::RIGHT:
+        return vertex.x() < vertex.w();
+    case ClipPlane::TOP:
+        return vertex.y() < vertex.w();
+    case ClipPlane::BOTTOM:
+        return vertex.y() > -vertex.w();
+    case ClipPlane::NEAR:
+        return vertex.z() > -vertex.w();
+    case ClipPlane::FAR:
+        return vertex.z() < vertex.w();
+    }
+
+    return false;
+}
+
+// TODO: This needs to interpolate color/UV data as well!
+FloatVector4 GL::Clipper::clip_intersection_point(const FloatVector4& vec, const FloatVector4& prev_vec, ClipPlane plane_index)
+{
+    // Uhhh, can somebody explain this math to me, please
+    FloatVector4 u, w;
+    FloatVector4 ret = prev_vec;
+    FloatVector4 plane = clip_planes[plane_index];
+    FloatVector4 plane_normal = clip_plane_normals[plane_index];
+
+    u = vec;
+    u -= prev_vec;
+    w = prev_vec;
+    w -= plane;
+    float d = plane_normal.dot(u);
+    float n = -plane_normal.dot(w);
+
+    ret += (u * (n / d));
+    return ret;
+}
+
+void GL::Clipper::clip_triangle_against_frustum(Vector<FloatVector4>& input_verts)
+{
+    Vector<FloatVector4, 6> clipped_polygon;
+    Vector<FloatVector4, 6> input = input_verts;
+    Vector<FloatVector4, 6>* current = &clipped_polygon;
+    Vector<FloatVector4, 6>* output_list = &input;
+    ScopeGuard guard { [&] { input_verts = *output_list; } };
+
+    for (u8 plane = 0; plane < NUMBER_OF_CLIPPING_PLANES; plane++) {
+        swap(current, output_list);
+        clipped_polygon.clear();
+
+        if (current->size() == 0) {
+            return;
+        }
+
+        // Save me, C++23
+        for (auto i = 0; i < static_cast<i32>(current->size()); i++) {
```

Comment:
```suggestion
        for (size_t i = 0; i < current->size(); i++) {
```

Reply:
What's the effect on the modulus below this? If `i` is `0`, then we'll underflow to `0xFFFFFFFF`.

Reply:
that wouldnt work with signed integers anyways, you would be trying to access a negative index of the vector, which would ASSERT

Reply:
Curious, it does appear that `(-1)%x` == `-1` in C++
I'm intrigued by why this doesn't crash

Reply:
> that wouldnt work with signed integers anyways, you would be trying to access a negative index of the vector, which would ASSERT

Interesting. This was a mistake with my and @alimpfard 's understanding with this. I'll get this fixed up.